lib/rules/utils/ast-utils.js

```diff
@@ -1357,17 +1357,43 @@ module.exports = {
      * next to each other, behavior is undefined (although it should return `true` in most cases).
      */
     canTokensBeAdjacent(leftValue, rightValue) {
+        const espreeOptions = { ecmaVersion: 2020, comment: true, range: true };
+
         let leftToken;
 
         if (typeof leftValue === "string") {
-            const leftTokens = espree.tokenize(leftValue, { ecmaVersion: 2015 });
+            const tokens = espree.tokenize(leftValue, espreeOptions);
```

Comment:
Do you mind adding a test for when a `Shebang` comment token is given as the left value? I don't think this could ever be a real world use case (since another token can't be on the same line as the shebang), but I do think we should make sure that this won't crash the process.

For reference, we mutate the text input to make the shebang a regular JS line comment before we parse in [`lib/linter/linter.js`](https://github.com/eslint/eslint/blob/master/lib/linter/linter.js#L632).

Reply:
You're right, that would crash! Maybe we could already add `try-catch` as suggested in [#12291 (comment)](https://github.com/eslint/eslint/issues/12291#issuecomment-552086069) and also close #12291 with this PR?

Reply:
`canTokensBeAdjacent` also didn't assume it could get a code (as a string argument) without any tokens, so it doesn't support a comment-only code. This isn't changed with this PR, it would still crash, so it isn't possible to call it with e.g.,:

```js
astUtils.canTokensBeAdjacent(someToken, "/* comment */");
```

Then, it might also make sense to support empty code as well, and maybe even to account for spaces at the start/end of the code (e.g., `canTokensBeAdjacent("a ", "b")` would return `true`).

Maybe we could do that as well in this PR?

Reply:
That sounds good to me! I guess another option (or something we could do additionally) is just do is short circuit it and add an early return of `false` if the left value token is of type `Shebang`. That would save the process from having to run the parse at all.

Reply:
> That sounds good to me! I guess another option (or something we could do additionally) is to short circuit it and add an early return of `false` if the left value token is of type `Shebang`. That would save the process from having to run the tokenizing at all.

Maybe I didn't understand, but I think it isn't possible to check if the token is `Shebang` before tokenizing, because we don't have the token yet?

Reply:
Oh, I see, I missed that it could be a string or a token. Yes, if the left value is a string that won't work. I think the approach you sounds good :+1: